python/pfs/pipe2d/generateCommands.py

```diff
@@ -0,0 +1,1493 @@
+
+        print(f"{shellCommand(command)}", file=fout)
+
+
+@export
+class CalibSource:
+    """Sources to construct a type of calib.
+    The concrete type of the calib is defined by subclasses.
+
+    Parameters
+    ----------
+    config : `CommandConfig`
+        Configurations used in constructing this calib.
+    source : `SourceFilter`
+        Sources for this calib.
+    validity : `int`
+        Valid days of the resulting calib.
+
+    Notes
+    -----
+    To make a subclass, ``typeName``` and ``commandName`` are required:
+
+        class DarkSource(
+                CalibSource,
+                typeName="dark", commandName="constructPfsDark.py"):
+            pass
+
+    ``typeName``
+        Name of the calib type. This is the name used in YAML files.
+    ``commandName``
+        Name of the command to construct the calib.
+    """
+
+    __slots__ = ["config", "source", "validity"]
+
+    # This property may be replaced by __subclasses__() once it is documented.
+    __subclasses: Dict[str, type] = {}
+
+    def __init__(self, config: CommandConfig, source: SourceFilter, validity: int):
+        self.config = config
+        self.source = source
+        self.validity = validity
+
+    def __init_subclass__(cls, *, typeName: str, commandName: str, **kwargs):
+        super().__init_subclass__(**kwargs)
+        CalibSource.__subclasses[typeName] = cls
+        cls.typeName = typeName
+        cls.commandName = commandName
+
+    @classmethod
+    def fromYaml(cls, yamlBlock: Mapping[str, Any]) -> "CalibSource":
+        """Construct ``CalibSource`` from a YAML block.
+
+        Parameters
+        ----------
+        yamlBlock : `Mapping[str, Any]`
+            A block of a YAML data structure:
+
+            ``"config"``
+                A config string, or a list of config strings.
+            ``"configfile"``
+                Path to a configuration file
+            ``"validity"``
+                Valid days for the resulting calib
+                before and after the date of the calib.
+            ``"id"``
+                A string like ``"field=BIAS"`` or a list of such strings.
+
+        Returns
+        -------
+        calibSource : `CalibSource`
+        """
+        config = CommandConfig.fromYaml(yamlBlock)
+        source = SourceFilter.fromYaml(yamlBlock)
+        validity = int(yamlBlock.get("validity", DEFAULT_CALIB_VALIDITY))
+        return cls(config, source, validity)
+
+    @staticmethod
+    def getSubclass(typeName: str) -> Type["CalibSource"]:
+        """Get a subclass by ``typeName``.
+
+        Parameters
+        ----------
+        typeName : `str`
+            ``typeName`` of the subclass.
+
+        Returns
+        -------
+        subclass : type
+            The class object of the subclass.
+        """
+        return CalibSource.__subclasses[typeName]
+
+    def execute(
+            self, logger: lsst.log.Log, fout: TextIO, dataDir: str, calib: str, rerun: str,
```

Comment:
All the methods take a `logger`. Maybe provide once to the constructor instead?

Reply:
Does it matter if I remove all `logger` parameters from `CalibSource` and `ScienceStep`? They are not used now. It may still be worth for `fromYaml()` method to have `logger` parameter.

Reply:
I don't have a strong opinion.